File: LSSEventListenerExample/src/main/java/com/espertech/QueriesDatabase/QueryMetadata.java
package com.espertech.QueriesDatabase;

import java.util.List;
import java.util.UUID;

public class QueryMetadata {
//    public QueryMetadata{
//        Objects.requireNonNull(name);
//        Objects.requireNonNull(deploymentId);
//        Objects.requireNonNull(query);
//        Objects.requireNonNull(eventClasses);
//        Objects.requireNonNull(category);
//    }
    public UUID id;
    public String name;
    public String queryStatement;
    public String deploymentId;
    public String query;
    public List<String> eventClasses;
    public String category;
    public Long createdAt;
    public Long updatedAt;
    public Boolean status;
    public String description;

    public QueryMetadata(String id, String name, String queryStatement, String deploymentId,
                         String query, List<String> eventClasses, String category, Long createdAt,
                         Long updatedAt, Boolean status, String description){

        this.id = UUID.fromString(id);
        this.name = name;
        this.queryStatement = queryStatement;
        this.deploymentId = deploymentId;
        this.query = query;
        this.eventClasses = eventClasses;
        this.category = category;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.status = status;
        this.description = description;
    }

    public QueryMetadata() {

    }

    public QueryMetadata(UUID id, String name, String queryStatement, String deploymentId,
                            String query, List<String> eventClasses, String category, Long createdAt,
                            Long updatedAt, Boolean status, String description) {
        this(id.toString(), name, queryStatement, deploymentId, query, eventClasses, category, createdAt, updatedAt, status, description);
    }

    @Override
    public String toString() {
        return "QueryMetadata{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", queryStatement='" + queryStatement + '\'' +
                ", deploymentId='" + deploymentId + '\'' +
                ", query='" + query + '\'' +
                ", eventClasses=" + eventClasses.stream().reduce((s1, s2) -> s1 + ", " + s2).orElse("No event classes") +
                ", category='" + category + '\'' +
                ", createdAt=" + createdAt +
                ", updatedAt=" + updatedAt +
                ", status=" + status +
                ", description='" + description + '\'' +
                '}';
    }
}

File: LSSEventListenerExample/UI/lss_thesis_ui/lib/Models/QueryMetadata.dart
class QueryMetadata {
  final String id;
  final String name;
  final String queryStatement;
  final String deploymentId;
  final String query;
  final String eventClasses;
  final String category;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool status;
  final String description;

  QueryMetadata({
    required this.id,
    required this.name,
    required this.queryStatement,
    required this.deploymentId,
    required this.query,
    required this.eventClasses,
    required this.category,
    required this.createdAt,
    required this.updatedAt,
    required this.status,
    required this.description,
  });

  QueryMetadata copyWith({
    String? query,
  }) {
    return QueryMetadata(
      id: id,
      name: name,
      queryStatement: queryStatement,
      deploymentId: deploymentId,
      query: query ?? this.query,
      eventClasses: eventClasses,
      category: category,
      createdAt: createdAt,
      updatedAt: updatedAt,
      status: status,
      description: description,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'queryStatement': queryStatement,
        'deploymentId': deploymentId,
        'query': query,
        'eventClasses': eventClasses.split(',').map((e) => e.trim()).toList(),
        'category': category,
        'createdAt': createdAt.millisecondsSinceEpoch,
        'updatedAt': updatedAt.millisecondsSinceEpoch,
        'status': status,
        'description': description,
      };

  factory QueryMetadata.fromJson(Map<String, dynamic> json) {
    return QueryMetadata(
      id: json['id'],
      name: json['name'],
      queryStatement: json['queryStatement'],
      deploymentId: json['deploymentId'],
      query: json['query'],
      eventClasses: (json['eventClasses'] as List<dynamic>).join(', '),
      category: json['category'],
      createdAt: DateTime.fromMillisecondsSinceEpoch(json['createdAt']),
      updatedAt: DateTime.fromMillisecondsSinceEpoch(json['updatedAt']),
      status: json['status'],
      description: json['description'],
    );
  }
}
